### ISSUE TITLE
Failed to select a proper baseline for training

### ISSUE THREAD

**Issue by SallyHuangImperial** [nianticlabs/monodepth2#314]
> So after applying the ratio of the baselines, I don't need extra scales for obtaining the metric depth, right?

That's correct!

> And also no matter what values I set for min_max distances, as long as I know the baseline, I will be able to obtain the metric depth by calculating (baseline * inverse depth from monodepth2) , right?

As long as the scaled (metric) min and max depth make sense given the scenes you are capturing, and that they do not make training unstable.

I hope this helps!

_Originally posted by @mrharicot in https://github.com/nianticlabs/monodepth2/issues/235#issuecomment-678549124_

**Comment by SallyHuangImperial** [comment]
Hi, I am still working in this field:)

Now I have a new dataset with ground truth, for which the depth range is between 5mm and 260mm and the baseline is about 4.9mm. Actually, if I use this code and set the baseline to 0.1 and the depth range to 5-260, I cannot obtain very accurate metric depth by multiplying 49. However, when I use the median or mean factor to multiply my depth output, the results make more sense. I want to ask how did you find the relationship between the hardcode baseline and the depth range when you were working with KITTI? And how did you find 0.1 as a proper baseline value for training?
I use the stereo mode to train my custom dataset. From the previous issues I know that if training with stereo mode, the metric depth can be obtained. I failed to recover the metric depth by just multiply 49 so I want to ask, is it OK if recover the scale by calculating the mean or median factor?
Thank you!

**Comment by mrharicot** [comment]
Hi,

The value of 0.1 for the hardcoded baseline was picked by trial and error to ensure stable initial convergence given the other hyperparameters.
Since monodepth2 predicts metric depth _up to scale_ we know we just have to find that single value and in the case of kitti we know that the actual baseline is 0.54m. A simple cross multiplication gives us that we need to scale our predictions by 5.4 to get meters.

If you train with stereo the proper way to get metric depth should be to use the baselines ratio, what average scale do you get when you use the median scaling?

**Comment by SallyHuangImperial** [comment]
Hi, 
Actually, the scale depends on the min/max range I set in the disp_to_depth function in the testing. I don't know if I should use meters or millimetres for the depth range. In the training procedure, I set the baseline to 0.1 and min_depth = 0.1, max_depth = 100 same as your settings. The real baseline of my dataset is 4.9mm and the depth range is 5-260mm. In the testing, I also used the disp_to_depth function to recover the depth. When the depth range is set to 0.1-100 in the testing, the scale is between 114 and 282, mean 140, different from frame to frame. When the depth range is set to 0.5-100 in the testing, the scale is between 23 and 57, mean 26. When the depth range is set to 0.5-260 in the testing, the scale is between 56 and 22, mean 28. When the depth range is set to 5-260 in the testing, the scale is between 2.3 and 6, mean 2.8. When the depth range is set to 35-260 in the setting，the scale is between 0.45 and 1.09, mean 0.56. 
I don't know if using the median scale to correct every frame makes sense. Do you have any suggestion?
Thank you!

**Comment by mrharicot** [comment]
If your baseline, min and max are 5, 5 and 260mm then training with 0.1, 0.1 and 100 would result in (relatively) a possible range of 5 to 500mm so you should be able to train using the default configuration.
You can then convert to metric depth by scaling the output by 49 to convert to millimeters.
Using median scaling is not a solution as it requires ground truth depth and is only used to measure accuracy of _relative or unscaled_ depth.
If you want metric depth you need to use a single scaling factor for all images.

**Comment by SallyHuangImperial** [comment]
Hi,

Thank you for the patient explanations. I agree with you that using median scaling does not make sense in my case. I have another question, when you used ( 0.1 - 100) in the function disp_to_depth, the depth result would be between 0.1 and 100. In the evaluate_depth.py, after multiplying 5.4, the result will be scaled to (0.54-540), right? I saw you use np.logical_and() to generate a mask between (0.001 - 80). I am confused here. How about the depth value between 80 and 540? And also why did you set depth range between 0.1 and 100 in the training even if you have a baseline at 0.1? If the depth range is set to (0.000185 - 19) and by multiplying 5.4, the metric depth could be (0.001 - 80). Much appreciate it if you could give some clues to me.
Thank you!

**Comment by mrharicot** [comment]
Hi,

We still want to be able to predict the correct value for such points so we need our maximum predicted value to be larger as there are pixels that are definitely much further than 80m in the data, they just happen not to have ground truth.
We mask out depths larger than 80m as there is no ground truth lidar depth past that value.
This was first used by Garg and we reused it to reproduce their evaluation. Note that this artificially lowers the error for such points.

**Comment by SallyHuangImperial** [comment]
Hi,

Thank you! I understood! And after mask out outliers I found that the depth range in my case is between 30 and 260mm. Do you think it will be more reasonable to change the depth range to 30 and 260 during the training? Actually when I modified the depth range, the results get worse. So I guess that the depth range of (0.1, 100) matches your code and your case so when they are changed, the training results are abnormal. What's your opinion?
Thank you!

**Comment by SallyHuangImperial** [comment]
Hi Godard,

Thank you for your reply!
Actually, no. Maybe I haven't found a proper baseline for the customer data
training.
Do you have any suggestions?

Clément Godard ***@***.***> 于2021年5月7日周五 下午10:03写道：

**Comment by mdfirman** [comment]
It might be worth doing some trial and error here using your datasets, trying out different depth ranges and baselines. For KITTI we certaintly didn't choose these values very scientifically, and so we don't have a great answer for how to select them for a new dataset.

But do make sure to monitor tensorboard closely early on in training, as the images there should be very helpful to tell if your training is likely to learn anything at all. (Try training a KITTI model to see what I mean)

I'm going to close this issue for now, but feel free to update on your progress here.
